Callback exceptions

bendk opened the pull request:
This is a work-in-progress for handling exceptions in callback interfaces.  It changes a few things:

- The `Throws` attribute is now offecially supported on CallbackInterface methods, allowing callback interfaces to specify which errors they will throw.  If we catch one of those exceptions, we serialize it to the `RustBuffer` that's normally used for the return value and send it back to Rust.
- On the Rust side, this means the callback interface trait methods need to change to return `Result<T>` instead of `T`.
- For now, `Throws` is optional and we fall back on the old behavior if it's not present.  But this is deprecated.
- We also handle unexpected exceptions in callback interface.  If we see one of those, we stringify it and serialize it to the `RustBuffer`.  This gets converted into a `uniffi::UnexpectedCallbackError` in Rust.  Users need to write a `From<>` impl for that struct to their error enum (unless they're on the old/deprecated behavior).
- I updated `callback_interfaces.md` to describe the new system, updated the callbacks example, and implemented it on Python/Kotlin.
- I made a slight change to the FFI return values for callback interface call.  See my [comment below](https://github.com/mozilla/uniffi-rs/pull/1344#discussion_r972681969) for details.

mhammond commented on examples/callbacks/src/lib.rs:
```diff
@@ -2,25 +2,33 @@
+    InternalTelephoneError,
+}
+
+impl From<uniffi::UnexpectedUniFFICallbackError> for TelephoneError {
```
I think a comment here - even just to point at the doc changes you made - might help people unfamiliar with uniffi and checking out the examples.

mhammond commented on uniffi/src/ffi/foreigncallbacks.rs:
```diff
@@ -131,8 +132,12 @@ use std::sync::atomic::{AtomicUsize, Ordering};
+/// * A callback returns:
+///    - `-2` An error occured that was serialized to buf_ptr
+///    - `-1` An unexpected error ocurred
+///    - `0` is a deprecated way to signal that if the call succeeded, but did not modify buf_ptr
```
So I *think* this means that even a function returning `Ok(())` will modify `buf_ptr`, so this is deprecated because it can only happen for a function which returns a void not in a Result - which is deprecated. If that's correct it might be worth expanding this (or not - I'm really just trying to understand how this is "deprecated" from the POV of the user)

mhammond commented on uniffi/src/ffi/foreigncallbacks.rs:
```diff
@@ -131,8 +132,12 @@ use std::sync::atomic::{AtomicUsize, Ordering};
+///    - `-2` An error occured that was serialized to buf_ptr
+///    - `-1` An unexpected error ocurred
+///    - `0` is a deprecated way to signal that if the call succeeded, but did not modify buf_ptr
+///    - `1` If the call succeeded.  For non-void functions the return value should be serialized
```
but this confuses me a little given the above - so maybe I'm wrong about what `Ok(())` does?

mhammond commented on uniffi_bindgen/src/bindings/kotlin/templates/CallbackInterfaceTemplate.kt:
```diff
@@ -34,11 +34,32 @@ internal class {{ foreign_callback }} : ForeignCallback {
+                    {%- endmatch %}
+                } catch (e: Throwable) {
+                    // Unexpected error
+                    -1
```
it's a little sad we can't optionally exact some data from the unexpected error if the bindings allow it. Or if we think that might be a possibility in the future, could the `Unexpected` error struct hold an `Option<String>` to make that future more tractable?

(I don't really care about that, just an idle thought)

mhammond commented on uniffi_bindgen/src/bindings/python/templates/ErrorTemplate.py:
```diff
@@ -1,19 +1,23 @@
 
+    {%- if e.is_flat() %}
+    class {{ variant_type_name }}({{ type_name }}):
+        def __unicode__(self):
```
I thought this was `__str__` in python 3? https://docs.python.org/3/reference/datamodel.html#object.__str__ doesn't mention `__unicode__` and a quick test:

```
class T:
#  def __str__(self):
#    return "str"
  def __unicode__(self):
    return "uni"

t = T()
print(t)
```

As shown, that prints `<__main__.T object at 0x00000156BC9CA140>` but if `__str__` is uncommented, `str` is printed

mhammond reviewed (commented):
I like this alot! I skimmed over some of the implementation details but think this is great and definitely what we should pursue.

bendk commented:
I added Swift support and also support for serializing error strings for unexpected errors.  I'm thinking this is pretty much what we want.

There's one last issue.  This change requires implementing `FfiConverter.read()` in Rust for errors.  However, this is not backwards compatible, since we currently allow errors to always be specified as "flat" even if they have fields.  Todolist is an example of this.  This works fine now, since we never need to construct the error in Rust, but once we try to implement `read()` we get compile errors.  Even worse, Todolist uses this with tuple-style fields which we don't have a syntax to support and I'm guessing it's not the only crate to do this.  I think there are a couple ways forward, which path do others think is best?

- Forcing users to switch to the fielded-style of errors, which probably means we need to add support for tuple-style fields.
- Only try to implement the `read()` method if the error is used in a callback interface and trying to solve the problem later.

bendk commented:
After thinking about this one more, I definitely think the second bullet is correct for now at least.  It's pretty easy to do and there's no need to tie this PR to deprecating an error style that I believe we currently use in production.  I updated the code, added a note about this in the suggested doc changes here, and made #1348 for the future work.

I removed the WIP label I think this one is ready for final review.

mhammond commented on docs/manual/src/udl/callback_interfaces.md:
```diff
@@ -18,28 +18,56 @@ by the host operating system (e.g. the key chain).
+In order to support errors in callback interfaces, UniFFI must be able to
+properly [lift the error](../internals/lifting_and_lowering.md).  This means
+that the if the error is described by an `enum` in the UDL it must be truely
+fieldless --  all variants must be unit variants.
```
I don't understand this para - (a) you seem to be saying that if a callback interface throws an error, all variants of that error must be fieldless, but that's not correct is it? I'd expect maybe just the "unexpected" variant should be fieldless? (b) but even that's not true - below you have a callback interface throwing `UnexpectedErrorWithReason` and that takes a string.

ISTM the error variant can be anything at all, but the practical limitation is just that it must be createable given just a `uniffi::UnexpectedUniFFICallbackError`?

(Actually, now I get into the impl, I think I better understand - "truely" is doing a lot of heavy lifting here :) So maybe reword that to better convey that the Rust implementation of the error must not have any fields. Looking at the implementation, it's also not clear what happens if that's not true - should we have a uitest for this scenario?

bendk replied:
I think you might have figured this out, but I wanted to write it out for posterity.  The issue is using the enum-style definition in the UDL, which doesn't list the fields, when in fact the variants do have fields.  It's now an issue because we have to generate read/lift methods for the scaffolding, so UniFFI needs to know how to construct the variants.

I updated the docs, hopefully it's a bit more clear.  A uitest is a great idea, I put one together and tried to make the error messages as obvious as a could -- although they're still not that obvious.  AFAIK, there's no way to do a static assertion for this and have it print out a nice error message.  If anyone knows one, please share.

mhammond commented on uniffi/src/ffi/foreigncallbacks.rs:
```diff
@@ -195,3 +200,26 @@ impl ForeignCallbackInternals {
+
+impl fmt::Display for UnexpectedUniFFICallbackError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "UnexpectedUniFFICallbackError(reason: {})", self.reason)
```
should reason be quoted here? Or probably just ":?".

mhammond commented on uniffi_bindgen/src/bindings/kotlin/templates/ErrorTemplate.kt:
```diff
@@ -76,12 +76,36 @@ public object {{ e|ffi_converter_name }} : FfiConverterRustBuffer<{{ type_name }
 
     @Suppress("UNUSED_PARAMETER")
```
here and below, I think the `@Suppress` annotations can now die?

bendk replied:
Yup.  Also, even though the param isn't used for flat enums the `@Supress` is still not needed since it's overriding an interface method.

mhammond commented on uniffi_bindgen/src/interface/mod.rs:
```diff
@@ -219,6 +219,25 @@ impl ComponentInterface {
         self.errors.iter().find(|e| e.name == name)
     }
 
+    /// Should we generate read/lower functions for errors?
```
This lower/read is from the POV of the foreign bindings, right? Maybe change this to start with "Should foreign bindings generate ..."?

mhammond replied:
oh, no, it's for the scaffolding - so should "lower" be "lift"?

mhammond commented on uniffi_bindgen/src/interface/mod.rs:
```diff
@@ -219,6 +219,25 @@ impl ComponentInterface {
 
+    /// Should we generate read/lower functions for errors?
+    ///
+    /// This is a temporary measure since lower/write can't be generated for some errors,
```
This is still slightly confusing to me (which is a comment on me, not on this patch :) But in what way is this temporary?

bendk replied:
I think that was optimistic on my part.  Once we implement `#1348`, then we can tell users that they shouldn't use the `enum` style definitions for errors if their errors actually have fields, and at some point we can drop support for it.  Maybe a simpler road would be to fully implement proc-macros and drop support for UDL.  But both of those things seems far off so I changed the wording.

mhammond commented on uniffi_bindgen/src/scaffolding/templates/CallbackInterfaceTemplate.rs:
```diff
@@ -69,31 +71,82 @@ impl r#{{ trait_name }} for {{ trait_impl }} {
+                    {%- endmatch %}
+                }
+                // 0 is a deprecated method to indicates success for void returns
+                0 => {
```
IIUC, we only expect to see this for bindings outside of this crate (and maybe Ruby 😅 ) - so how about arranging for a deprecated warning here? (Deprecating and finally removing things without generating a warning seems very likely to catch people by surprise later)

bendk replied:
Good call.  I added an `eprintln!` for this, is that the best way to signal a warning?

mhammond replied:
A mildly better way might be to have a noop function marked `#[deprecated] ` and call it here, but I'm really not too bothered

mhammond reviewed (changes requested):
This is looking good to me. If we really are deprecating non-result functions in callbacks, I'd really like to see a deprecation warning issued (which probably means this should be a breaking change, which I think is fine) and *possibly* a uitest. OTOH, I guess we could soften this down to a recommendation (ie, "if you want to avoid panics, you should..."). TBH I don't care too much as I hope we can deprecate callback interfaces entirely later, but it's too soon for me to be confident that's going to be possible.

(I'll also try and come back to this to better understand the semantics around error fields, but I don't see that blocking anything here)
